### DISCUSSION TITLE
Kiai time offset can sometimes appear to be off by 1ms on compatibility-exported lazer beatmaps

### DISCUSSION THREAD

**Discussion by felipemarins** [ppy/osu#30595] [Q&A] [answered]
As the title says, sometimes, when you export a lazer beatmap to .osz, the kiai time may be off by 1ms. I suppose this is due to timing point's precision that is not supported by stable.

Here are some beatmap files where it occurs (one is the .osz, and the other is a .olz for comparison):
[beatmaps-for-testing.zip](https://github.com/user-attachments/files/17727361/beatmaps-for-testing.zip)

Edit: It seems that it happens to hit objects too (but it doesn't trigger AiMod in this case).

**Comment by smoogipoo** [comment]
I do wonder if we're doing wrong by truncating rather than rounding... @OliBomby is there a reason for truncating specifically?

**Reply by OliBomby** [reply]
The reason we're truncating is because osu stable truncates instead of rounding for any float values in the .osu that have to be int. Also snapping to timing ticks in the timeline in stable truncates to integer. If any hitobject or timing point were to be rounded up while the uninherited point doesn't move it would be 1 ms off.

Idk yet what the problem here could be, I would need to look into it.

**Comment by OliBomby** [comment, marked as answer]
The problem seems to be that we are truncating the offset of the uninherited timing point without taking into account that it changes the snapping for all objects that come after it. This isn't fixed by rounding instead of truncating.

We'll need to offset all the objects in the timing section by the offset due to truncating the uninherited timing point, and then truncate the objects themselves. For example if the uninherited point is at 1000.4 ms and an object is at 2000.2 ms, we need to first move the object back 0.4 ms to 1999.8 ms and then truncate it to 1999 ms.
